src/pattern_matching/pssm/mod.rs

```diff
@@ -0,0 +1,154 @@
+// except according to those terms.
+
+//! Implementation of position-specific scoring matrix (PSSM), aka position
+//! weight matrix (PWM), for both DNA and amino-acid sequences
+
+use ndarray::prelude::Array2;
+use std::f32::NEG_INFINITY;
+use utils::IntoTextIterator;
+
+mod dnamotif;
+mod protmotif;
+
+pub use self::dnamotif::DNAMotif;
+pub use self::protmotif::ProtMotif;
+
+pub const EPSILON: f32 = 1e-5;
+pub const INVALID_MONO: u8 = 255;
+
+/// represents motif score
+#[derive(Debug, Clone)]
+pub struct ScoredPos {
+    pub loc: usize,
+    pub sum: f32,
+    pub scores: Vec<f32>,
+}
+
+impl Default for ScoredPos {
+    fn default() -> ScoredPos {
+        ScoredPos {
+            loc: 0,
+            sum: NEG_INFINITY,
+            scores: Vec::new(),
+        }
+    }
+}
+
+pub trait Motif {
+    /// lookup table mapping monomer -> index
+    const LK: [u8; 127] = [INVALID_MONO; 127];
+    const MONOS: &'static [u8] = b"";
+
+    /// use lk to find index; enforce boundaries
+    fn lookup(mono: u8) -> Option<usize> {
+        if mono >= 127 {
+            None
+        } else {
+            let idx = Self::LK[mono as usize];
+            if idx == INVALID_MONO {
+                None
+            } else {
+                Some(idx as usize)
+            }
+        }
+    }
+    /// reverse lookup: given index, return monomer (eg, 0 -> A)
+    fn rev_lk(idx: usize) -> u8;
+
+    fn len(&self) -> usize;
+
+    /// represent highly conserved bases
+    fn degenerate_consensus(&self) -> Vec<u8>;
+    fn get_scores(&self) -> &Array2<f32>;
+    /// sum of "worst" base at each position
+    fn get_min_score(&self) -> f32;
+    /// sum of "best" base at each position
+    fn get_max_score(&self) -> f32;
+    // this should be replaced with a CTFE ... or maybe just a constant
+    fn get_bits() -> f32;
+
+    // standard PSSM scoring is calibrated to (1) pattern len and (2) the min and
+    //     max possible scores
+    // this is just a dumb sum of matching bases
+    fn raw_score<'a, T: IntoTextIterator<'a>>(&self, seq_it: T) -> (usize, f32, Vec<f32>) {
+        let pwm_len = self.len();
+
+        let mut best_start = 0;
+        let mut best_score = -1.0;
+        let mut best_m = Vec::new();
+        // we have to look at slices, so a simple iterator won't do
+        let seq = seq_it.into_iter().cloned().collect::<Vec<u8>>();
+        let scores = self.get_scores();
+        for start in 0..seq.len() - pwm_len + 1 {
+            let m: Vec<f32> = (0..pwm_len)
+                .map(|i| scores[[i, Self::lookup(seq[start + i]).expect("raw lookup")]])
+                .collect();
+            let tot = m.iter().sum();
+            if tot > best_score {
+                best_score = tot;
+                best_start = start;
+                best_m = m;
+            }
+        }
+        (best_start, best_score, best_m)
+    }
+
+    /// apply PSM to sequence, finding the offset with the highest score
```

Comment:
This should be PSSM, not PSM? In general, could this PR you just one name, either PSSM or PWM? Or is it necessay to distinguish?